pkg/cloud/aliyunprovider.go

```diff
@@ -0,0 +1,828 @@
+			log.Infof("Pricing information for node with same features %s already exists hence skipping", lookupKey)
+			continue
+		}
+		// TO-DO: Check if Node pricing already available , skip it if available
+		if client, ok = aliyun.clients[slimK8sNode.RegionID]; !ok {
+			client, err = sdk.NewClientWithAccessKey(slimK8sNode.RegionID, aak.AccessKeyId, aak.AccessKeySecret)
+			if err != nil {
+				return fmt.Errorf("access key provided does not have access to location %s", slimK8sNode.RegionID)
+			}
+			aliyun.clients[slimK8sNode.RegionID] = client
+		}
+		signer = signers.NewAccessKeySigner(aak)
+		pricingObj, err = processDescribePriceAndCreateAliyunPricing(client, slimK8sNode, signer, c)
+
+		if err != nil {
+			return err
+		}
+		aliyun.Pricing[lookupKey] = pricingObj
+	}
+
+	// TO-DO: PV pricing
+	// //get pvList ultimately from aliyun cloud provider and resemble data from the pvtype to
+	// // Hardcodedk8sNodeDiskStruct
+	// pvList := aliyun.Clientset.GetAllPersistentVolumes()
+
+	// pvList := []*Hardcodedk8sNodeDiskStruct{}
+	// pvList = append(pvList, &Hardcodedk8sNodeDiskStruct{
+	// 	DiskType:         "data",
+	// 	DiskCategory:     "cloud",
+	// 	PerformanceLevel: "",
+	// 	RegionID:         "cn-hangzhou",
+	// 	PriceUnit:        "Hour",
+	// 	SizeInGiB:        60,
+	// 	ProviderID:       "Ali-XXX-pv-01",
+	// }, &Hardcodedk8sNodeDiskStruct{
+	// 	DiskType:         "data",
+	// 	DiskCategory:     "cloud",
+	// 	PerformanceLevel: "P1",
+	// 	RegionID:         "cn-hangzhou",
+	// 	PriceUnit:        "Hour",
+	// 	SizeInGiB:        40,
+	// 	ProviderID:       "Ali-XXX-pv-01",
+	// })
+
+	// for _, pv := range pvList {
+	// 	if client, ok = aliyun.clients[pv.RegionID]; !ok {
+	// 		client, err = sdk.NewClientWithAccessKey(pv.RegionID, aak.AccessKeyId, aak.AccessKeySecret)
+	// 		if err != nil {
+	// 			return fmt.Errorf("access key provided does not have access to location %s", pv.RegionID)
+	// 		}
+	// 		aliyun.clients[pv.RegionID] = client
+	// 	}
+	// 	signer = signers.NewAccessKeySigner(aak)
+	// 	pricingObj, err = processDescribePriceAndCreateAliyunPricing(client, pv, signer)
+	// 	lookupKey, err = determineKeyForPricing(pv)
+	// 	if err != nil {
+	// 		return err
+	// 	}
+	// 	aliyun.Pricing[lookupKey] = pricingObj
+	// }
+	// log.Infof("Length of pricing is %d", len(aliyun.Pricing))
+	// log.Infof("random value is %v", aliyun.Pricing[lookupKey])
+	return nil
+}
+
+// AllNodePricing returns all the billing data for nodes and pvs
+func (aliyun *Aliyun) AllNodePricing() (interface{}, error) {
+	aliyun.DownloadPricingDataLock.RLock()
+	defer aliyun.DownloadPricingDataLock.RUnlock()
+	return aliyun.Pricing, nil
+}
+
+// NodePricing gives a specific node for the key
+func (aliyun *Aliyun) NodePricing(key Key) (*Node, error) {
+	aliyun.DownloadPricingDataLock.RLock()
+	defer aliyun.DownloadPricingDataLock.RUnlock()
+
+	// Get node features for the key
+	keyFeature := key.Features()
+
+	pricing, ok := aliyun.Pricing[keyFeature]
```

Comment:
What is a "feature"? Is there a better name?

Reply:
Key Interface has feature [method](https://github.com/opencost/opencost/blob/develop/pkg/cloud/provider.go#L120), ideally its something used to get the information for node/pv matching this feature.